File: 904-fruit-into-baskets/904-fruit-into-baskets.cpp
#include <algorithm>
#include <unordered_map>

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int i = 0;
        int max_so_far = INT_MIN;
        unordered_map<int,int> map;
        
        for (int j = 0; j < fruits.size(); j++) {
            map[fruits[j]]++;
            
            if (map.size() > 2) {
                map[fruits[i]]--;
                if (map[fruits[i]] == 0) map.erase(fruits[i]);
                i++;
            }
            
            max_so_far = max(max_so_far, j - i + 1);
        }
        
        return max_so_far;
    }
};

File: 904-fruit-into-baskets/904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        res = 0
        
        l = 0
        basket = {}
        
        for r in range (len(fruits)):
            basket[fruits[r]] = 1 + basket.get(fruits[r], 0)
            
            while len(basket) > 2:
                basket[fruits[l]] -= 1
                if basket[fruits[l]] == 0: 
                    del basket[fruits[l]]
                l+=1
            res = max(res, r-l+1)
            
        return res